src/features/runtime-checks.js

```diff
@@ -0,0 +1,249 @@
+            return listener[0] !== args[0] || listener[1] !== args[1]
+        })
+    }
+
+    toString () {
+        // TODO
+        let interfaceName = ''
+        if (this._tagName === 'script') {
+            interfaceName = 'Script'
+        }
+        return `[object HTML${interfaceName}Element]`
+    }
+
+    get tagName () {
+        return this.#tagName.toUpperCase()
+    }
+
+    get nodeName () {
+        return this.tagName
+    }
+
+    remove () {
+        const el = this.getElement()
+        if (el) {
+            return el.remove()
+        }
+        return super.remove()
+    }
+
+    removeChild (child) {
+        const el = this.getElement()
+        if (el) {
+            return el.removeChild(child)
+        }
+        return super.removeChild(child)
+    }
+}
+
+/**
+ * Overrides the instanceof checks to make the custom element interface pass an instanceof check
+ * @param {Object} elementInterface
+ */
+function overloadInstanceOfChecks (elementInterface) {
+    const proxy = new Proxy(elementInterface[Symbol.hasInstance], {
+        apply (fn, scope, args) {
+            if (args[0] instanceof DDGRuntimeChecks) {
+                return true
+            }
+            return Reflect.apply(fn, scope, args)
+        }
+    })
+    Object.defineProperty(elementInterface, Symbol.hasInstance, {
+        value: proxy
+    })
+}
+
+/**
+ * Returns true if the tag should be intercepted
+ * @param {string} tagName
+ * @returns {boolean}
+ */
+function shouldInterrogate (tagName) {
+    const interestingTags = ['script']
+    if (!interestingTags.includes(tagName)) {
+        return false
+    }
+    const stack = getStack()
+    const scriptOrigins = [...getStackTraceOrigins(stack)]
+    const isInterestingHost = scriptOrigins.some(origin => {
+        return interestingHosts.find(hostname => matchHostname(origin, hostname))
+    })
+    return isInterestingHost
+}
+
+function overrideCreateElement () {
+    const proxy = new DDGProxy(featureName, document, 'createElement', {
+        apply (fn, scope, args) {
+            if (args.length >= 1) {
+                const initialTagName = args[0].toLowerCase()
+                if (shouldInterrogate(initialTagName)) {
+                    args[0] = 'ddg-runtime-checks'
+                    const el = Reflect.apply(fn, scope, args)
+                    el.setTagName(initialTagName)
+                    return el
+                }
+            }
+            return Reflect.apply(fn, scope, args)
+        }
+    })
+    proxy.overload()
+    initialCreateElement = proxy._native
+}
+
+export function load () {
+    customElements.define('ddg-runtime-checks', DDGRuntimeChecks)
+}
+
+export function init (args) {
+    const domain = args.site.domain
+    const domains = getFeatureSetting(featureName, args, 'domains') || []
+    const enabled = domains.find((rule) => {
+        if (rule.matchAll) {
+            return true
+        }
```

Comment:
Under the `domains` setting, there is a list of `rule`s. Isn't this the wrong way around?

Also, if one rule contains `matchAll: true`, then it will nullify all the rest of the rules, right? Why not make that a top-level setting?

Reply:
The structure is:

```
    "settings": {
        "domains": [
            {
                "domain": "mozilla.org"
            }
        ]
    }
```

This is to allow progressive rollout of the feature, I take your point I could make a `matchAllDomains` key though.

The intention here is to delete this after initial rollout.